src/components/mortgage-investment-compare/input-card.js

```diff
@@ -18,328 +19,374 @@ import Accordion from '../general/accordion'
 import Tooltip from '../general/tooltip'
 import ValidatedNumberInput from '../general/validated-number-input'
 
-const InputCard = props => {
-  function onInputChange (path, value) {
-    const newInput = _.cloneDeep(props.input)
+class InputCard extends React.Component {
+  constructor () {
+    super()
+    this.onInputChange = this.onInputChange.bind(this)
+  }
+
+  componentWillReceiveProps (nextProps) {
+    if (
+      this.props.input !== nextProps.input ||
+      this.props.includeROI !== nextProps.includeROI ||
+      this.props.includeInflation !== nextProps.includeInflation
+    ) {
+      this.props.updtateResult(nextProps.input)
+    }
+  }
```

Comment:
Why is this needed?

Reply:
@hdavidzhu it seems like there are 4-5 questions/comments about this so I will try to address all of them here and explain the thought process.

What's happening now is that we are adding 2 new properties to the state: `includeROI` and `includeInflation`. We need to listen for changes in these and update the results every time the user enables/disables inflation. This has always been done with lifecycle methods in React, and now more recently with hooks.

1. Hooks: hooks are great but changing `MortgageInvestmentCompare` to a function would have required significant work so I decided to not do it this way.

2. Lifecycle methods: historically, `componentDidUpdate` has always been the default lifecycle method to monitor changes in the state within a component. I could have used `componentDidUpdate` in `MortgageInvestmentCompare` to monitor changes in `includeROI`  and `includeInflation`, but `componentDidUpdate` is now deprecated. So I decided to use `componentWillReceiveProps` instead which monitors changes in the props being passed to the child components. 

I also initially decided to call `getResults` within `setState`'s callback as explained above, but that's a very "dirty and quick" way of doing it.

Because we are now monitoring for changes in the parent's state within the child components, we need to pass in a function that changes the state in the parent whenever changes are detected. As you said, `getResults` was purse, so calling this in the child component and setting state there wouldn't have worked as it would set the state on the child component as well. While your concerns about the function becoming stateful now make sense, before we were setting the state in the `updateInput` handler like so:

```
updateInput (input) {
    const result = this.getResult(input)
    this.setState({ input, result })
  }
```

Which isn't really a great approach since the goal of `updateInput` is to update the input, and I don't find it particularly dangerous to have an `updateResult`.  What we do with `updateResult` right now is no different from what we do with `updateInput` as `updateInput` also gets called within the child component through the onChange handlers, and runs in the parent to change the state. Passing more responsibility and information to the child components is completely fine and normal.  If you'd prefer it, we could go back to using `getResult` to keep it pure, and then we can create a new method `updateResult` which only sets the state with whatever `getResult` returns. 

The current approach is probably not the cleanest and perhaps not as straightforward for someone that hasn't used React for some time, but it's safe and gets the job done. In the future, we should consider using the new hooks which would have worked great in this situation. Converting everything we have right now to use hooks instead wouldn't be that hard but is unnecessary at this point.

Reply:
> What's different now is that results are not affected only by changes in the input's on change handler so this also doesn't get the job done.

What if you handled that in the parent component by updating the results when you detect changes in the input?

> Converting everything we have right now to use hooks wouldn't be that hard but is unnecessary at this point.

I think it's pretty straightforward to get away w/o using it. Since pairing is tricky for us, I'll spin up a temporary PR to demonstrate what I mean

Reply:
@jgozal Here's what I had in mind: https://github.com/jgozal/moneycompass/pull/62

Give it a run and lemme know what you think

Reply:
@hdavidzhu here it is https://github.com/jgozal/moneycompass/pull/61/commits/239f358a8fad676ed6ccf7e013fa2cb1625f05fc

Making another small commit to fix old docs